megablocks/layers/mlp.py

```diff
@@ -153,17 +154,43 @@ class MemoryOptimizedMLP(torch.autograd.Function):
 
     @staticmethod
     @torch.cuda.amp.custom_fwd
-    def forward(ctx, x, w1, w2, topo):
+    def forward(ctx, x, w1, w2, topo, num_bits):
         # x: [m, k], w1: [n, k], w2: [n, k]
         if (not x.is_contiguous() or not w1.is_contiguous() or
             not w2.is_contiguous()):
             raise ValueError("Expected contiguous 'x', 'w1' and 'w2'.")
+        if num_bits not in (-1, 4, 8):
+            raise ValueError("Activation quantization num_bits must be " +
+                             f"-1 (disabled), 4, or 8. Got {num_bits}")
+
+        topo_tensors = (topo.row_indices,
+                        topo.column_indices,
+                        topo.offsets,
+                        topo.column_indices_t,
+                        topo.offsets_t,
+                        topo.block_offsets_t)
 
         # Layer 0: x @ w1.t().
         sdd_out = stk.ops.sdd(x, w1.t(), topo)
 
         # GeLU.
-        gelu_out = gelu.gelu(sdd_out)
+        if num_bits == -1:
+            gelu_out = gelu.gelu(sdd_out)
+            input_save_args = (x, sdd_out.data)
+        else:
+            # safe version: allocate new mem for the gelu output
+            hidden_q, hidden_scales, gelu_out_data = turbo.quantize_signed(
+                sdd_out.data, num_bits=num_bits, op=turbo.ElemwiseOps.GELU_FORWARD)
+            gelu_out = stk.Matrix(sdd_out.shape, gelu_out_data, *topo_tensors)
+
+            # # bold version: lie about __restrict__ and do the GELU in-place
```

Comment:
Fwiw, I lie about __restrict__ all the time. You just have to make sure your kernel maintains correctness if the pointers are equal, since the compiler won't help you. If you do this in turbo, you could expose an 'in-place' API as well so that the caller doesn't have to worry about the 'boldness'.